src/haz3lcore/zipper/ProjectorBase.re

```diff
@@ -0,0 +1,175 @@
+ * position and dimensions in base editor could be
+ * added here if needed */
+[@deriving (show({with_path: false}), sexp, yojson)]
+type info = {
+  id: Id.t,
+  syntax,
+  ci: option(Info.t),
+};
+
+let info_init = (p: syntax) => {id: Piece.id(p), syntax: p, ci: None};
+
+/* To add a new projector, implement this module signature */
+module type Projector = {
+  [@deriving (show({with_path: false}), sexp, yojson)]
+  type model;
+  /* The internal model type of the projector which will
+   * be serialized and persisted. Use `unit` if you don't
+   * need other state beyond the underlying syntax */
+  [@deriving (show({with_path: false}), sexp, yojson)]
+  type action;
+  /* An internal action type to be used in actions which
+   * update the model. Use `unit` if the basic projector
+   * actions (type `action`) above suffice */
+  let init: model;
+  /* Initial state of the model */
+  let can_project: Piece.t => bool;
+  /* A predicate determining if the given underlying
+   * syntax (currently limited to convex pieces) is
+   * supported by this projector. This is used to gate
+   * adding the projector */
+  let can_focus: bool;
+  /* Does this projector have internal position states,
+   * overriding the editor caret & keyboard handlers?
+   * If yes, the focus method will be called when this
+   * projector is either clicked on or if left/right
+   * is pressed when the caret is to the immediate
+   * right/left of the projector */
+  let view:
+    (
+      model,
+      ~info: info,
+      ~local: action => Ui_effect.t(unit),
+      ~parent: external_action => Ui_effect.t(unit)
+    ) =>
+    Node.t;
+  /* Renders a DOM view for the projector, given the
+   * model, an info packet (see info type for details),
+   * and has two callbacks: ~parent for parent editor
+   * actions(see external_action type above), and ~local
+   * for this projector's local update function. */
+  let placeholder: (model, info) => shape;
+  /* How much space should be left in the code view for
+   * this projector? This determines how the base code
+   * view is laid out, including how movement around the
+   * projector works. In principle this could be derived
+   * from the view, but this is awkward to do so for now
+   * projector writers are responsible for keeping these
+   * in sync with each other. */
+  let update: (model, action) => model;
+  /* Update the local projector model given an action */
+  let focus: ((Id.t, option(Direction.t))) => unit;
+  /* Does whatever needs to be done to give a projector
+   * keyboard focus. Right now this is only for side
+   * effects but could be extended in the future to
+   * take/return the model if the projector needs to
+   * maintain a complex internal position state */
+};
+
+/* Projector model and action are serialized so that
+ * they may be used by the Editor without it having
+ * specialized knowledge of projector internals */
+type serialized_model = string;
+type serialized_action = string;
```

Comment:
I would like to talk more about this and see if there's some way we can hide the internals from the `Editor` while still having the rich types internal to the projector.

Reply:
that actually is done here. see the ProjectorBase.Cook functor